### DISCUSSION TITLE
SGP40 Air Quality Sensor on D1 Mini with Tasmota 13.3.0(sensors)

### DISCUSSION THREAD

**Discussion by TMHEINE** [arendst/Tasmota#20597] [Support request]
Hi,
how can i get the SGP40-Sensor work on this standart sensor built?
It is connected like here:
![image](https://github.com/arendst/Tasmota/assets/6831714/3d7baaa7-6e9b-4a47-8cef-cdb8a9558732)
And Sensor was found:
![image](https://github.com/arendst/Tasmota/assets/6831714/11841672-47f6-4df0-b391-2dc2b24469c3)

What do I have to do to get the sensor information on the main screen?

![image](https://github.com/arendst/Tasmota/assets/6831714/257267fd-fccb-4dba-b122-4f44ee9ac565)

Thanks in advance

**Comment by sfromis** [comment]
Check if your `I2CDriver` command output indicates that index 69 included and enabled. If the sensor is detected (instead of just present on the bus), a message like this should appear shortly after booting.
`00:00:00.134 I2C: BMP280 found at 0x76`
(Obviously, my example is with a different sensor)

**Reply by TMHEINE** [reply]
I2CDriver 69 is active But nothing found...
![image](https://github.com/arendst/Tasmota/assets/6831714/f8ee61c5-3854-4ac3-9242-5d4c6463bb77)

**Reply by sfromis** [reply]
Well, did you see the detection message after boot?

Anyway, Tasmota has drivers for both SGP4X and SGP40. While I'm not familiar with the reasoning or differences here, maybe `USE_SGP4X` (I2C driver index 82) would be relevant, with a custom build (this is just speculation on my part).

**Reply by TMHEINE** [reply]
Hello,
I uas using it in version 12.3.1 with custom image.
![image](https://github.com/arendst/Tasmota/assets/6831714/e31f35a0-af5d-43cc-9511-f9eeafcd98f5)
But I thought it is now implementet in the actual Sensors-version

**Reply by sfromis** [reply]
You did not answer about seeing (or not) the detection message right after boot. Was your old 12.3.1 build with `USE_SGP40` or `USE_SGP4X`? And yes, the former should be included in `tasmota-sensors.bin.gz`, not the latter.

**Reply by TMHEINE** [reply]
No, there is nothing during startup:
![image](https://github.com/arendst/Tasmota/assets/6831714/cba322e3-7cdb-4156-b4ee-41716bf1967c)

**Reply by TMHEINE** [reply]
What is the difference between USE_SGP40 or USE_SGP4X? I have a SGP40.
Ordered here:
https://de.aliexpress.com/item/1005004680000599.html

**Reply by TMHEINE** [reply]
I ordered an other SPG40 from a different company. Same result. 
It is not possible do get a SPG40 or SPG41 running withe the Standard-Sensor-Image.
Even in Version 13.4.0.2
Or did I make something wrong?

**Reply by barbudor** [reply]
What is the result of command 'i2cdriver'?
You should have the number 69 without a `!`

**Reply by TMHEINE** [reply]
It is there, for sure without "!"

![image](https://github.com/arendst/Tasmota/assets/6831714/3875e5f1-583c-4e78-b0c6-99452cddb0da)
![image](https://github.com/arendst/Tasmota/assets/6831714/49f64840-2eeb-4a2b-99fa-944152df4f95)

**Reply by TMHEINE** [reply]
Does anyone have an SGP40 sensor with the standard sensor firmware in use?

**Comment by TMHEINE** [comment]
I have now generated a firmware with "#define USE_SGP40 // [I2cDriver69] Enable SGP40 sensor (I2C address 0x59) (+1k4 code)".
The SGP40 was not found.
Then I generated a firmware with "#define USE_SGP4X // [I2cDriver82] Enable SGP41 sensor (I2C address 0x59) (+7k2 code)".
The SGP40 was then found after deactivating the I2CDriver69.
Unfortunately the displayed data 
![image](https://github.com/arendst/Tasmota/assets/6831714/d6a15666-2d6b-4c8e-b7c3-244fac33e522)
does not match the SGP40 . It should look like this:
![image](https://github.com/arendst/Tasmota/assets/6831714/7cdb26a8-1fdf-4a83-ba53-923bd5bce7af)

If I then connect an SGP41 here, it is recognized correctly:
![image](https://github.com/arendst/Tasmota/assets/6831714/743e69a3-a2c4-4344-8bd0-ab20207e2579)

So in my opinion there is an fault in programm-code. 
Maybe someone can find it...

**Reply by barbudor** [reply]
The 2 drivers present the results in 2 different ways
So don't expect the SGP4X driver to present the results the same way as the SGP40 driver

What I can see is the SGP4X driver seems to work correctly with both your SGP40 and your SGP41 as the values are very close (assuming the same room at short time interval)

So at least you have a working solution with the SGP4X driver

Without having a SGP40 in hand it's hard to debug the SGP40 driver
I see that the code doesn't have much error logging
I could build a test version for you to test to try to understand where the SGP40 driver fails

**Reply by barbudor** [reply]
In the attached ZIP, htere is a tasmota-sensors.bin.gz which have more logs in hte SGP40 driver
It will help to understand why the SGP40 driver doesn't find the device

[tasmota-sensors.bin.zip](https://github.com/arendst/Tasmota/files/14487444/tasmota-sensors.bin.zip)

Please provide the logs BY COPY/PASTE using triple-backticks - NO SCREEN SHOTS OF LOGS
Like :
![image](https://github.com/arendst/Tasmota/assets/2996042/38de6e0b-bb6f-4c49-aac8-f1e971b3aab4)

**Reply by TMHEINE** [reply]
Hello,
here the Result of the Log:

```00:00:00.002 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F7, Count 13
00:00:00.056 QPC: Count 1
00:00:00.058 CFG: CR 378/699, Busy 0
00:00:00.063 TYA: Active=0
00:00:00.071 ROT: Mode 1
00:00:00.073 SRC: Restart
00:00:00.075 Project tasmota - Tasmota Version 13.4.0.2(sensors)-2_7_6(2024-03-04T22:37:58)
00:00:00.088 SHT: Sensor did not ACK command
00:00:00.106 SGP40: check I2C 0x59
00:00:00.107 SGP40: call begin
00:00:00.108 SGP40.begin: new Adafruit_I2CDevice
00:00:00.120 SGP40.begin: SerialNumber = 0x00 467 6CF1
00:00:00.132 SGP40.begin: featureset = 0x3240
00:00:00.383 SGP40.selfTest: read failed
00:00:00.435 WIF: Checking connection...
00:00:00.436 WIF: Attempting connection...
00:00:01.886 WIF: Connecting to AP1 TMHEINE Channel 1 BSSId 74:42:7F:9F:A8:E8 in mode 11n as tasmota-DBCE1F-3615...
00:00:03.405 WIF: Checking connection...
00:00:03.406 WIF: Attempting connection...
00:00:04.407 WIF: Checking connection...
00:00:04.408 WIF: Connected
00:00:04.660 HTP: Web server active on tasmota-DBCE1F-3615 with IP address 192.168.178.10
00:00:04.907 WIF: Sending Gratuitous ARP
00:00:04.909 NTP: Sync time...
00:00:04.940 WIF: DNS resolved '2.pool.ntp.org' (195.201.137.97) in 29 ms
00:00:04.973 RTC: UTC 2024-03-05T05:23:42Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
06:23:42.001 RTC: Synced by NTP
06:23:42.693 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"13.4.0.2(sensors)","FallbackTopic":"cmnd/DVES_DBCE1F_fb/","GroupTopic":"cmnd/tasmotas/"}}
06:23:42.698 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-DBCE1F-3615","IPAddress":"192.168.178.10"}}
06:23:42.708 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":5}}
06:23:44.940 QPC: Reset
06:23:45.016 HTP: Main Menu
06:23:46.905 APP: Boot Count 5
06:23:46.910 RSL: STATE = {"Time":"2024-03-05T06:23:46","Uptime":"0T00:00:09","UptimeSec":9,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"TMHEINE","BSSId":"74:42:7F:9F:A8:E8","Channel":1,"Mode":"11n","RSSI":98,"Signal":-51,"LinkCount":1,"Downtime":"0T00:00:04"}}
06:23:47.195 CFG: Saved to flash at F6, Count 14, Bytes 4096
06:23:51.894 RSL: config = {"ip":"192.168.178.10","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-DBCE1F-3615","mac":"E8DB84DBCE1F","md":"Generic","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"13.4.0.2","t":"tasmota_DBCE1F","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1}
06:23:51.965 RSL: sensors = {"sn":{"Time":"2024-03-05T06:23:51"},"ver":1}
06:23:55.130 HTP: Console
06:24:01.423 WIF: Checking connection...
06:24:21.441 WIF: Checking connection...
06:24:37.911 WIF: Sending Gratuitous ARP
06:24:39.660 CMD: i2cscan
06:24:39.662 SRC: WebConsole from 192.168.178.39
06:24:39.664 CMD: Grp 0, Cmd 'I2CSCAN', Idx 1, Len 0, Pld -99, Data ''
06:24:39.683 RSL: RESULT = {"I2CScan":"Device(s) found at 0x59"}
06:24:41.424 WIF: Checking connection...
06:24:50.722 CMD: i2cdriver
06:24:50.724 SRC: WebConsole from 192.168.178.39
06:24:50.726 CMD: Grp 0, Cmd 'I2CDRIVER', Idx 1, Len 0, Pld -99, Data ''
06:24:50.732 RSL: RESULT = {"I2CDriver":"7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,46,48,69,76"}```

**Reply by TMHEINE** [reply]
And here is the log of the other device (different manufacturer):
```00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F6, Count 14
00:00:00.055 QPC: Count 1
00:00:00.057 CFG: CR 378/699, Busy 0
00:00:00.062 TYA: Active=0
00:00:00.070 ROT: Mode 1
00:00:00.072 SRC: Restart
00:00:00.074 Project tasmota - Tasmota Version 13.4.0.2(sensors)-2_7_6(2024-03-04T22:37:58)
00:00:00.087 SHT: Sensor did not ACK command
00:00:00.106 SGP40: check I2C 0x59
00:00:00.106 SGP40: call begin
00:00:00.107 SGP40.begin: new Adafruit_I2CDevice
00:00:00.119 SGP40.begin: SerialNumber = 0x00 438 CCBD
00:00:00.131 SGP40.begin: featureset = 0x3240
00:00:00.382 SGP40.selfTest: read failed
00:00:00.434 WIF: Checking connection...
00:00:00.435 WIF: Attempting connection...
00:00:01.887 WIF: Connecting to AP1 TMHEINE Channel 1 BSSId 74:42:7F:9F:A8:E8 in mode 11n as tasmota-DBCE1F-3615...
00:00:03.402 WIF: Checking connection...
00:00:03.403 WIF: Connected
00:00:03.659 HTP: Web server active on tasmota-DBCE1F-3615 with IP address 192.168.178.10
00:00:03.907 WIF: Sending Gratuitous ARP
00:00:03.908 NTP: Sync time...
00:00:03.934 WIF: DNS resolved '2.pool.ntp.org' (80.153.195.191) in 24 ms
00:00:03.979 RTC: UTC 2024-03-05T05:26:23Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
06:26:23.000 RTC: Synced by NTP
06:26:23.700 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"13.4.0.2(sensors)","FallbackTopic":"cmnd/DVES_DBCE1F_fb/","GroupTopic":"cmnd/tasmotas/"}}
06:26:23.705 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-DBCE1F-3615","IPAddress":"192.168.178.10"}}
06:26:23.715 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":6}}
06:26:24.605 HTP: Main Menu
06:26:26.165 HTP: Console
06:26:26.902 QPC: Reset
06:26:27.880 RSL: STATE = {"Time":"2024-03-05T06:26:27","Uptime":"0T00:00:08","UptimeSec":8,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"TMHEINE","BSSId":"74:42:7F:9F:A8:E8","Channel":1,"Mode":"11n","RSSI":94,"Signal":-53,"LinkCount":1,"Downtime":"0T00:00:03"}}
06:26:28.869 APP: Boot Count 6
06:26:29.212 CFG: Saved to flash at F5, Count 15, Bytes 4096
06:26:32.889 RSL: config = {"ip":"192.168.178.10","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-DBCE1F-3615","mac":"E8DB84DBCE1F","md":"Generic","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"13.4.0.2","t":"tasmota_DBCE1F","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1}
06:26:32.959 RSL: sensors = {"sn":{"Time":"2024-03-05T06:26:32"},"ver":1}
06:26:42.438 WIF: Checking connection...
06:27:02.438 WIF: Checking connection...
06:27:08.673 CMD: i2cscan
06:27:08.675 SRC: WebConsole from 192.168.178.39
06:27:08.677 CMD: Grp 0, Cmd 'I2CSCAN', Idx 1, Len 0, Pld -99, Data ''
06:27:08.696 RSL: RESULT = {"I2CScan":"Device(s) found at 0x59"}```

**Reply by barbudor** [reply]
Thanks
I see it fails in the selfTest as the device doesn't respond to the read

I will check possible cause by reading the code and SGP doc

**Reply by barbudor** [reply]
I see that there are design flaws in the original SGP40 driver that has been addressed in the SGP4X driver
Considering that the SGP4X driver works for both SGP40 and SGP41 I don't see a point in trying to fix the SGP40 driver which would require a lot of rewrite

Do you see any interrest in that old driver ?
If there are missing feature in the SGP4X driver, it sounds better to port them to SGP4X driver than trying to maintain the SGP40 driver

**Reply by sfromis** [reply]
This thread from last summer suggests that what became the SGP4X driver had trouble handling both SGP40 and SGP41. https://github.com/arendst/Tasmota/discussions/18867

**Reply by TMHEINE** [reply]
I think it's best to keep an eye on the SGP4x driver as it can actually handle both.
But I have more questions: 
-Is it possible to read the temperature and humidity of an SGP40/41?
-SGP40driver is not working anymore, can the SGP4xdriver be placed in the standard sensor-image

**Reply by sfromis** [reply]
Looks like neither SGP40 nor SGP41 can measure temp/hum, but they can try to adjust the calculations based on the resistance of heated and corroded tin if the driver can supply them with these values.

**Reply by TMHEINE** [reply]
In my old Project with the SGP40Driver the Abs humidity was shown.
See:
![image](https://github.com/arendst/Tasmota/assets/6831714/1bd9b96e-7ae7-42f0-af1c-2be3c6e05c14)
Where is this vaue from?

**Reply by sfromis** [reply]
That is a calculation when the driver supplied temp/hum, via the Tasmota global temp/hum values, in this case based on the BME280. It can *really* only measure resistance of heated and corroded tin, nothing more. All else is calculations, based on many assumptions which may not match very well always.

**Reply by TMHEINE** [reply]
Ok, then it is good for me.
Will the SGP4x driver be used in the Standard-Sensor Image in Future?
